content/advent-2019/flags.md

```diff
@@ -0,0 +1,279 @@
+	flag.StringVar(&config.host, "host", config.host, "host to listen on") // [3]
+	flag.Usage = func() {                                                  // [4]
+		fmt.Fprintf(flag.CommandLine.Output(), usage, os.Args[0])
+		flag.PrintDefaults()
+	}
+	flag.Parse() // [5]
+
+	http.HandleFunc("/", handler)
+	addr := fmt.Sprintf("%s:%d", config.host, config.port)
+	fmt.Printf("server ready on %s\n", addr)
+	if err := http.ListenAndServe(addr, nil); err != nil {
+		log.Fatalf("error: %s", err)
+	}
+
+}
+
+func init() { // [6]
+	// Set defaults
+	s := os.Getenv("HTTPD_PORT")
+	p, err := strconv.Atoi(s)
+	if err == nil {
+		config.port = p
+	} else {
+		config.port = 8080
+	}
+
+	h := os.Getenv("HTTPD_HOST")
+	if len(h) > 0 {
+		config.host = h
+	} else {
+		config.host = "localhost"
+	}
+}
+
+func handler(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Hello Gophers\n")
+}
+```
+
+1. I tend to use a `config` struct for configuration instead of separate
+  variables. When applications evolve, the number of configuration option will
+  grows and I'd like to keep them in one place
+2. `flag.IntVar` will bind `config.port` to the `-port` command line option
+3. `flag.StringVar` will bind `config.host` to the `-host` command line option
+4. Set `flag.Usage` to a function that will print your help
+5. `flag.Parse` will parse command line arguments and will print help when
+  calling your application with `-h` or `--help`. `flag.Parse` will exit the
+  program on any command line error
+6. You can use `init` to set default values and populate values from environment
+  variables
+
+## Validation
+
+A good practice is to validate all the command line switches at program start.
+The `flag` packages have built in function for integers, floats, boolean,
+time.Duration and more. However sometimes you'd like to have your own type.
+Using `flag.Var` we can achieve this.
+
+We'll define `portVar` struct that will implement the
+[flag.Value](https://golang.org/pkg/flag/#Value) interface. We'll also provide
+a `PortVar` function to create such a variable.
+
+Then we'll change our main to use `PortVar` instead of `IntVar`.
+
+```go
+func main() {
+	flag.Var(PortVar(&config.port), "port", "port to listen on")
+	// ...
+}
+
+func PortVar(port *int) *portVar {
+	return &portVar{port}
+}
+
+type portVar struct {
+	port *int
+}
+
+func (p *portVar) String() string {
+	if p.port == nil {
+		return ""
+	}
+
+	return fmt.Sprintf("%d", *p.port)
+}
+
+func (p *portVar) Set(s string) error {
+	val, err := strconv.Atoi(s)
+	if err != nil {
+		return err
+	}
+
+	const minPort, maxPort = 1, 65535
+	if val < minPort || val > maxPort {
+		return fmt.Errorf("port %d out of range [%d:%d]", val, minPort, maxPort)
+	}
+
+	*p.port = val
+	return nil
+}
+```
+
+## Sub Commands
+
+Instead of having one executable to start the HTTP server and another to check
+it's alive. We can have one executable that does both commands (same as `git`
```

Comment:
This seems like it should be a comma connecting the two sentences?

Reply:
Yup. Fixed.